### PR TITLE
chore: improve `polycli fund`

### PR DESCRIPTION
**Pull request by leovct**
# Description

- [x] chore: migrate from `web3` library to `go-ethereum`.
- [x] feat: Implement a `Funder` smart contract to be able to bulk fund wallets in one transaction. It speeds up the funding process by a lot. 🚀
- [x] chore: refactor flags, check flags in `PreRunE` and move flag logic to `app`.
- [x] feat: enable to fund wallets specified by user
- [x] feat: enable to derive wallets from private key
- [x] chore: use `RunE` instead of `RunE` when running the `fund` command.
- [x] chore: update `--eth-amount` in `polycli loadtest`. Users can now specify the amount in eth instead of having to specify the hexadecimal representation of the amount in wei.

## Jira / Linear Tickets

- [DVT-1114](https://polygon.atlassian.net/browse/DVT-1114) - migrate to `go-ethereum` lib.
- [DVT-1116](https://polygon.atlassian.net/browse/DVT-1116) - enable to fund wallets specified by user.
- [DVT-1115](https://polygon.atlassian.net/browse/DVT-1115) - enable to fund wallets derived from private key.

# Testing

```bash
# Fund wallets specified by the user.
$ polycli fund --addresses=0x5eD3BE7a1cDafd558F88a673345889dC75837aA2,0x1Ec6efdBd371D6444779eAE7B7e16907e0c8eC27
3:58PM INF Starting bulk funding wallets
3:58PM INF Using addresses provided by the user
3:58PM INF Wallet(s) funded! 💸
3:58PM INF Total execution time: 1.020693583s

# Fund 20 random wallets using a pre-deployed contract address.
$ polycli fund --number=20 --contract-address=0xf5a73e7cfcc83b7e8ce2e17eb44f050e8071ee60
3:58PM INF Starting bulk funding wallets
3:58PM INF Deriving wallets from the default mnemonic
3:58PM INF Wallet(s) derived count=20
3:58PM INF Wallet(s) funded! 💸
3:58PM INF Total execution time: 396.814917ms

# Fund 20 random wallets.
$ polycli fund --number 20 --hd-derivation=false
3:58PM INF Starting bulk funding wallets
3:58PM INF Generating random wallets
3:58PM INF Wallet(s) generated count=20
3:58PM INF Wallets' address(es) and private key(s) saved to file fileName=wallets.json
3:58PM INF Wallet(s) funded! 💸
3:58PM INF Total execution time: 1.027506s
```

Extract from `wallets.json`.

```json
[
  {
    "Address": "0xc1A44B1e37EE1fca4C6Fd5562c730d5b8525e4C6",
    "PrivateKey": "c1a8f737fd9f78aee361bfd856f9b2e99f853a5fe5efa2131fb030acdcee762b"
  },
  {
    "Address": "0x5D8121cf716B70d3e345adB58157752304eED5C3",
    "PrivateKey": "fbc57de542cef10fdcdf99e5578ffb5508992e9a8623ea4a39ab957d77e9b849"
  },
  ...
]
```

Check the balances of the wallets.

```bash
$ cast balance 0xc1A44B1e37EE1fca4C6Fd5562c730d5b8525e4C6
50000000000000000

$ cast balance 0x5D8121cf716B70d3e345adB58157752304eED5C3
50000000000000000
...
```

Send funds on behalf on one of the funded account.

```bash
$ cast send --private-key e211490bcaae916e6cdbf58877e7fc697a51c6e11cfe591b0b2b72e6123e4d12 --value 0.01ether 0x469838206407A001B7086F102fe35b62967170d3

blockHash               0xd9dbd1b377d75cb3704c2e1ca27b3054e18afe692abed8eb621a4a231446adfb
blockNumber             8
contractAddress         
cumulativeGasUsed       21000
effectiveGasPrice       3401058594
gasUsed                 21000
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
root                    
status                  1
transactionHash         0x530d68a9f705e02fcf02362e1e75cdeb76e49f2bd9df165f6c447df512652367
transactionIndex        0
type                    2
```

[DVT-1114]: https://polygon.atlassian.net/browse/DVT-1114?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DVT-1116]: https://polygon.atlassian.net/browse/DVT-1116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DVT-1115]: https://polygon.atlassian.net/browse/DVT-1115?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/fund/cmd.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+func init() {
+	p := new(cmdFundParams)
+	flagSet := FundCmd.Flags()
+
+	p.RpcUrl = flagSet.StringP("rpc-url", "r", "http://localhost:8545", "The RPC endpoint url")
+	p.PrivateKey = flagSet.String("private-key", defaultPrivateKey, "The hex encoded private key that we'll use to send transactions")
+
+	// Wallet parameters.
+	p.WalletCount = flagSet.Uint64P("wallets", "w", 10, "The number of wallets to fund")
+	p.WalletFundingHexAmount = flagSet.StringP("amount", "a", "0xb1a2bc2ec50000", "The amount of wei to send to each wallet") // 0xb1a2bc2ec50000 represensts 0.05 ether.
```
Comment (rebelArtists) [inline cmd/fund/cmd.go:55]: nit: just my opinion, but this change feels less intuitive for users. in my mind, user will want to specify amt in eth as human readable format for quicker iteration

**Review by rebelArtists — DISMISSED** [review]
nice updates, thanks Leo! looks like you may need to fix 1 failing ci job but otherwise lgtm

File: cmd/fund/cmd.go (hunk 1 above)
Comment (leovct) [reply cmd/fund/cmd.go:55]: Yep, I wanted to have the same behaviour on `polycli fund` and `polycli loadtest` but you're right, it's easier to use amount in eth format. Will update this asap.

Comment (leovct) [reply cmd/fund/cmd.go]: It's better like this :)

```bash
$ go run main.go fund --help | grep eth-amount
  -a, --eth-amount float        The amount of ether to send to each wallet (default 0.05)

$ go run main.go loadtest --help | grep eth-amount
      --eth-amount float                       The amount of ether to send on every transaction (default 0.001)
```

File: cmd/fund/usage.md [hunk 2]
```diff
@@ -1,10 +1,51 @@
...
\ No newline at end of file
+# Fund wallets specified by the user.
+$ polycli fund --addresses 0x5eD3BE7a1cDafd558F88a673345889dC75837aA2,0x1Ec6efdBd371D6444779eAE7B7e16907e0c8eC27
+3:43PM INF Starting bulk funding wallets
+3:43PM INF Using addresses provided by the user
+3:43PM INF Wallet(s) funded! 💸
+3:43PM INF Total execution time: 1.020693583s
+
+# Fund 20 wallets, derived from the default private key.
+$ polycli fund --number 20 --hd-derivation true
```
Comment (rebelArtists) [inline cmd/fund/usage.md:12]: awesome simplicity for user here!

File: cmd/fund/usage.md [hunk 3]
```diff
@@ -1,10 +1,51 @@
...
+3:44PM INF Starting bulk funding wallets
+3:44PM INF Deriving addresses from the default mnemonic
+Error: todo: implement this mode
```
Comment (rebelArtists) [inline cmd/fund/usage.md:15]: do we have follow-on ticket for this piece?

File: doc/polycli_loadtest.md [hunk 4]
```diff
@@ -114,6 +114,7 @@ The codebase has a contract that used for load testing. It's written in Yul and
   -c, --concurrency int                        Number of requests to perform concurrently. Default is one request at a time. (default 1)
       --erc20-address string                   The address of a pre-deployed ERC20 contract
       --erc721-address string                  The address of a pre-deployed ERC721 contract
+      --eth-amount float                       The amount of ether to send on every transaction (default 0.001)
```
Comment (rebelArtists) [inline doc/polycli_loadtest.md:117]: nice

File: doc/polycli_fund.md [hunk 5]
```diff
@@ -11,35 +11,78 @@
...
-      --funding-wallet-pk string   Corresponding private key for funding wallet address, ensure you remove leading 0x
-  -h, --help                       help for fund
-      --output-file string         Specify the output JSON file name (default "funded_wallets.json")
-      --rpc-url string             The RPC endpoint url (default "http://localhost:8545")
-      --wallet-count int           Number of wallets to fund (default 2)
-      --wallet-funding-amt float   Amount to fund each wallet with (default 0.05)
-      --wallet-funding-gas uint    Gas for each wallet funding transaction (default 100000)
+      --addresses strings         Comma-separated list of wallet addresses to fund
+      --contract-address string   The address of a pre-deployed Funder contract
+  -a, --eth-amount float          The amount of ether to send to each wallet (default 0.05)
```
Comment (rebelArtists) [inline doc/polycli_fund.md:79]: find this more intuitive, thanks!!

**Review by rebelArtists — APPROVED** [review]

File: cmd/fund/usage.md (hunk 3 above)
Comment (leovct) [reply cmd/fund/usage.md:15]: I already implemented yet, I just need to update `usage.md`!

File: cmd/fund/fund.go [hunk 6]
```diff
@@ -1,312 +1,262 @@
...
-		if err != nil {
-			log.Error().Err(err).Msg("Error getting nonce")
-			return 0, err
-		}
-		globalNonce = nonce
-		nonceInitialized = true
+// Initialize  parameters.
+func initializeParams(ctx context.Context, c *ethclient.Client) (*ecdsa.PrivateKey, *big.Int, error) {
+	// Parse the private key.
+	privateKey, err := crypto.HexToECDSA(strings.TrimPrefix(*params.PrivateKey, "0x"))
```
Comment (IdrisHanafi) [inline cmd/fund/fund.go:107]: nice, just added this to `polycli loadtest` too. given that the `--private-key` flag is used in a lot of place, the logic of trimming the `0x` will be reused a lot and we may have to eventually think of a approach to streamline in (maybe putting it in the `util` package, idk yet)

File: contracts/src/funder/Funder.sol
Comment (IdrisHanafi) [inline contracts/src/funder/Funder.sol:1]: clever approach

File: util/util.go [hunk 7]
```diff
@@ -203,3 +206,13 @@ func convHexToUint64(hexString string) (uint64, error) {
 	}
 	return uint64(result), nil
 }
+
+// BlockUntilSuccessfulFn is designed to wait until a specified number of Ethereum blocks have been
+// mined, periodically checking for the completion of a given function within each block interval.
+type BlockUntilSuccessfulFn func(ctx context.Context, c *ethclient.Client, f func() error) error
+
+func BlockUntilSuccessful(ctx context.Context, c *ethclient.Client, retryable func() error) error {
```
Comment (IdrisHanafi) [inline util/util.go:214]: hopefully this is a permanent home for this fxn. It has been moved around so many times hahaha

**Review by IdrisHanafi — APPROVED** [review]
nice job, lgtm 🔥 

tried testing it out with the sample commands in the PR description:
```
polycli fund --wallets 20 --verbosity 500
```

but realized, the flag has been changed from `wallets` to `number` lol.

File: util/util.go (hunk 7 above)
Comment (leovct) [reply util/util.go:214]: Not sure, I don't really like `util/util.go` lol but I didn't want to refactor too much 🥶

**Pull request merged**